dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
-	// separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// constraint.
+// Used inside a ResourceClaimSpecAlternatives or a ResourceClaimSpecification object.
+type ResourceClaimSpec struct {
+	// ControllerName defines the name of the resource driver that is
+	// meant to handle allocation of this claim. If empty,
+	// structured parameters are used for allocation.
+	//
+	// Resource drivers have a unique name in forward domain order
+	// (acme.example.com).
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	ControllerName string `json:"controllername,omitempty"`
+
+	// Config defines configuration parameters that apply to the entire claim.
+	// They are ignored while allocating the claim.
 	//
 	// +optional
 	// +listType=atomic
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,4,opt,name=config"`
+
+	// All constraints must be satisfied when allocating the claim.
+	Constraints []Constraint `json:"constraints"`
 
-	// Devices contains the actual device claim details, each entry
-	// of which must be sastisfied by selecting one or more devices.
-	// Each entry is either a claim for set of devices, or a list of
-	// prioritized claims for sets of devices, one of which must be
-	// satisfied.
+	// Requests are individual requests for separate resources for the claim.
+	// An empty list is valid and means that the claim can always be allocated
+	// without needing anything. A class can be referenced to use the default
+	// requests from that class.
 	//
-	// +required
 	// +listType=atomic
-	Devices []DeviceClaimInstance `json:"devices,omitempty"`
+	Requests []ResourceRequest `json:"requests,omitempty" protobuf:"bytes,5,name=requests"`
+
+	// Future extension, ignored by older schedulers. This is fine because scoring
+	// allows users to define a preference, without making it a hard requirement.
+	//
+	//
+	// Score *SomeScoringStruct
+
+	// Shareable indicates whether the allocated claim is meant to be shareable
+	// by multiple consumers at the same time.
+	// +optional
+	Shareable bool `json:"shareable,omitempty" protobuf:"bytes,3,opt,name=shareable"`
 }
 
-// DeviceClaimInstance captures a group of claims which must all be satisfied,
-// and a group for which one must be sastisfied. If you require multiple
-// groups, one of each may be satsfied, you will need multiple instances.
-type DeviceClaimInstance struct {
-	// Name is used to associate this instance with a specific
-	// container.
-	Name string `json:"name"`
+// Constraint must have one and only one field set.
+type Constraint struct {
+	// The match criteria must be satisfied before a set of devices will be used
+	// together.
+	//
+	// +optional
+	Match *MatchModel `json:"match,omitempty"`
```

Comment:
To answer this we need to decide if we ever envision other types of constraints which can co-exist with a match constraint. if the answer is **no**, then it makes sense to pull them up here and make constraints itself a one-of. If the answer is **yes**, then we should leave it as-is, so that `MatchModel` can be a one-of separate from `Constraints` (which would not be a one-of).

Reply:
I removed `MatchModel` and instead added `MatchAttribute` to `Contraint`.